data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java

```diff
@@ -13,21 +13,43 @@
 import javax.inject.Named;
 import java.io.Serializable;
 import java.util.Map;
+import java.util.List;
+import java.util.ArrayList;
 import java.util.function.Function;
 
 @Named
 class ParseTreeCoercionService {
     private final Map<Class<? extends Serializable>, Function<Object, Object>> literalTypeConversions;
+    private ExpressionFunctionProvider expressionFunctionProvider;
 
     @Inject
-    public ParseTreeCoercionService(final Map<Class<? extends Serializable>, Function<Object, Object>> literalTypeConversions) {
+    public ParseTreeCoercionService(final Map<Class<? extends Serializable>, Function<Object, Object>> literalTypeConversions, ExpressionFunctionProvider expressionFunctionProvider) {
         this.literalTypeConversions = literalTypeConversions;
+        this.expressionFunctionProvider = expressionFunctionProvider;
     }
 
     public Object coercePrimaryTerminalNode(final TerminalNode node, final Event event) {
         final int nodeType = node.getSymbol().getType();
         final String nodeStringValue = node.getText();
         switch (nodeType) {
+            case DataPrepperExpressionParser.Function:
+                final int funcNameIndex = nodeStringValue.indexOf("(");
+                final String functionName = nodeStringValue.substring(0, funcNameIndex);
+                final int argsEndIndex = nodeStringValue.indexOf(")", funcNameIndex);
+                final String argsStr = nodeStringValue.substring(funcNameIndex+1, argsEndIndex);
+                final String[] args = argsStr.split(",");
+                List<Object> argList = new ArrayList<>();
+                for (final String arg: args) {
+                    String trimmedArg = arg.trim();
+                    if (trimmedArg.charAt(0) == '/') {
```

Comment:
OK. I see now that this is where you resolve the JSON Pointer.

I think we should let the function decide which arguments are JSON Pointers and which are not. For example, the `length` function should state that the first argument is a JSON Pointer.

Also, we generally support a simpler JSON Pointer syntax without the leading slash if you are getting from the root object. So, this logic will not work for this situation.

Also, see that for CIDR checks, we want to include arguments which are not part of the event. https://github.com/opensearch-project/data-prepper/issues/2625#issuecomment-1532182607

Reply:
Can you elaborate more on what is "simpler JSON Pointer"? It looks like grammar only supports JSON pointer with forward slash.

Regarding CIDR checks, Could you provide some example usecases? Are you referring to issues like #2625? To any function that may be dealing with CIDR blocks, it would get a string argument like `192.168.10.*` and it is up to the function to convert it to network addresses. The grammar that I defined in this PR, only allows string or JsonPointer arguments. JsonPointer argument will be looked in the event and the resultant value can be of any "type". That's why the argument list is a list of Objects and not list of Strings

Reply:
On second thoughts, I think we need to pass `event` to `evaluate` because we may have to use the string arguments as an argument to event tags or event metadata but resolving JsonPointer in ParseTreeCoercionService still makes sense because it avoids code duplication.

Reply:
> Can you elaborate more on what is "simpler JSON Pointer"? It looks like grammar only supports JSON pointer with forward slash.

Yes, that is what I meant. I'm sorry about the confusion. I know we support this in the Event model (see this [test case](https://github.com/opensearch-project/data-prepper/blob/38b578ce37ae1efe20eae6b06546ce35bde1bd49/data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java#L285-L291)), but perhaps not in the grammar.

-----

I think that the function itself needs to know what its arguments are and how to interpret them. 

Consider the possibility of a `hasSubstring` method defined as `hasSubstring(eventPointer, substring)`. 

We can't be sure how pipeline authors would use it. Perhaps they have the following:

```
hasSubstring("/file", "/usr/var/linux/path");
```

This is a reasonable scenario. However, the current code would interpret that the second argument should be a JSON Pointer.

----

> On second thoughts, I think we need to pass event to evaluate because we may have to use the string arguments as an argument to event tags or event metadata

This is a good point. Indeed, `getTags()` will have not arguments (at least as initially planned).

> but resolving JsonPointer in ParseTreeCoercionService still makes sense because it avoids code duplication.

I think we should look to see how we can generalize `resolveJsonPointerValue`. One simple approach could be to also pass a `BiFunction<String, Event, Object>` into the `evaluate` method.

```
return expressionFunctionProvider.provideFunction(functionName, event, argList, this::resolveJsonPointerValue)
```

Another option is an `Event` decorator so that we could call `Event::get`, but that might be a larger undertaking.